### PR TITLE
feat: add additional operations to `BaseFlags`

### PR DESCRIPTION
**Pull request by celsiusnarhwal**
## Summary

This pull request implements addition (+), subtraction (-), union (|), intersection (&), and inversion (~) operations on the `BaseFlags` class and accordingly updates the documentation for its inheritors. The addition and union operations are equivalent and their operators interchangeable.

I made this change with the `Permissions` class in mind, intending to enable functionality such as adding two `Permissions` objects together to obtain the combined set of permissions they represent; subtracting two `Permissions` objects to set all permissions that are `True` in one object to `False` in the other; finding the intersection of two `Permissions` objects to obtain the set of permissions that are `True` in both; and inverting a `Permissions` object to set all of its `True` permissions to `False` and vice versa. However, this PR enables these operations on _all_ inheritors of `BaseFlags` - the previously-described actions could be similarly applied to objects of the `Intents` class, for example. 

The addition, subtraction, union, and intersection operations can operate on a `BaseFlags` object and a) another `BaseFlags` object, or b) a `flag_value` object. Reverse methods for each of these operations are also implemented, allowing operations between `BaseFlags` and `flag_value` objects to be properly commutative.

## Information

- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting, examples, ...).

## Checklist

- [x] I have searched the open pull requests for duplicates.
- [x] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why.

### CODE REVIEW

**Comment by EmreTech** [comment]
Considered adding this to all flags in `flag.py`?

File: discord/permissions.py [hunk 1]
```diff
@@ -149,11 +161,58 @@ def is_strict_superset(self, other: Permissions) -> bool:
...
+        try:
+            if isinstance(other, Permissions or int):
+                return self & ~other
+            elif isinstance(other, flag_value):
+                return self & ~other.flag
+        except TypeError:
+            raise TypeError(f"'-' not supported between instances of '{type(self)}' and '{type(other)}'")
+
+    def __invert__(self):
+        if isinstance(self, Permissions):
```
Comment (plun1331) [inline discord/permissions.py:200]: Unnecessary

File: discord/permissions.py [hunk 2]
```diff
@@ -149,11 +161,58 @@ def is_strict_superset(self, other: Permissions) -> bool:
...
+            return self.__class__(~self.value)
+        else:
+            raise TypeError(f"'~' not supported for '{type(self)}'")
+
+    __rand__ = __and__
+    __ror__ = __or__
+    __radd__ = __add__
+    __rsub__ = __sub__
```
Comment (plun1331) [inline discord/permissions.py:208]: These should be typehinted probably

File: discord/permissions.py [hunk 3]
```diff
@@ -149,11 +161,58 @@ def is_strict_superset(self, other: Permissions) -> bool:
...
     __le__: Callable[[Permissions], bool] = is_subset
     __ge__: Callable[[Permissions], bool] = is_superset
     __lt__: Callable[[Permissions], bool] = is_strict_subset
     __gt__: Callable[[Permissions], bool] = is_strict_superset
 
+
```
Comment (plun1331) [inline discord/permissions.py:214]: These extra newlines shouldn't be here

**Review by plun1331 — CHANGES_REQUESTED** [review]
While this is a nice feature, implementing it in the [`BaseFlags`](https://github.com/Pycord-Development/pycord/blob/1bb5ac38d20ebb6a84eb348cb2b6edb4d6e2f4cc/discord/flags.py#L102) class would probably be a better solution, also allowing you to perform the same operations on everything that inherits from it. (intents, etc)

**Comment by EmreTech** [comment]
> While this is a nice feature, implementing it in the [`BaseFlags`](https://github.com/Pycord-Development/pycord/blob/1bb5ac38d20ebb6a84eb348cb2b6edb4d6e2f4cc/discord/flags.py#L102) class would probably be a better solution, also allowing you to perform the same operations on everything that inherits from it. (intents, etc)

Yeah this is what I was trying to say

**Review by Lulalaby — DISMISSED** [review]
That looks like a mess

**Comment by celsiusnarhwal** [comment]
> While this is a nice feature, implementing it in the [`BaseFlags`](https://github.com/Pycord-Development/pycord/blob/1bb5ac38d20ebb6a84eb348cb2b6edb4d6e2f4cc/discord/flags.py#L102) class would probably be a better solution, also allowing you to perform the same operations on everything that inherits from it. (intents, etc)

That actually is a better idea; I made this PR with only the `Permissions` class in mind, so that did not initially occur to me. Nevertheless, I have moved my code changes to the `BaseFlags` class (with typehints this time) and updated the documentation for its inheritors accordingly.

File: discord/flags.py [hunk 4]
```diff
@@ -39,6 +39,7 @@
     overload,
 )
 
+import discord
```
Comment (EmreTech) [inline discord/flags.py:42]: Why are you directly importing the `discord` namespace? Just import what you need

File: discord/flags.py [hunk 5]
```diff
@@ -140,6 +141,44 @@ def __iter__(self) -> Iterator[Tuple[str, bool]]:
...
+        else:
+            raise TypeError(f"'-' not supported between instances of {type(self)} and {type(other)}")
+
+    def __invert__(self):
+        return self.__class__._from_value(~self.value)
+
+    __rand__: Callable[[BaseFlags], bool] = __and__
+    __ror__: Callable[[BaseFlags], bool] = __or__
+    __radd__: Callable[[BaseFlags], bool] = __add__
+    __rsub__: Callable[[BaseFlags], bool] = __sub__
```
Comment (EmreTech) [inline discord/flags.py:180]: I'm a little confused with these type annotations. The `__and__` and `__or__` functions support BaseFlags and flag_value for the `other` parameter

File: discord/flags.py [hunk 6]
```diff
@@ -177,6 +216,19 @@ class SystemChannelFlags(BaseFlags):
                Returns an iterator of ``(name, value)`` pairs. This allows it
                to be, for example, constructed as a dict or a list of pairs.
 
+            Adds two flags together.
```
Comment (EmreTech) [inline discord/flags.py:219]: The comment should go below the `.. describe::` part

File: discord/flags.py [hunk 7]
```diff
@@ -250,6 +302,19 @@ class MessageFlags(BaseFlags):
...
```
Comment (EmreTech) [inline discord/flags.py:305]: See above

File: discord/flags.py [hunk 8]
```diff
@@ -490,6 +568,19 @@ class Intents(BaseFlags):
...
```
Comment (EmreTech) [inline discord/flags.py:571]: See above

File: discord/flags.py [hunk 9]
```diff
@@ -1030,6 +1121,18 @@ class MemberCacheFlags(BaseFlags):
...
```
Comment (EmreTech) [inline discord/flags.py:1124]: See above

File: discord/flags.py [hunk 10]
```diff
@@ -1155,6 +1258,18 @@ class ApplicationFlags(BaseFlags):
             to be, for example, constructed as a dict or a list of pairs.
             Note that aliases are not shown.
 
+            Adds two flags together.
```
Comment (EmreTech) [inline discord/flags.py:1261]: See above

File: discord/flags.py [hunk 11]
```diff
@@ -1253,6 +1368,18 @@ class ChannelFlags(BaseFlags):
...
```
Comment (EmreTech) [inline discord/flags.py:1371]: See above

File: discord/permissions.py [hunk 12]
```diff
@@ -97,6 +97,18 @@ class Permissions(BaseFlags):
              Checks if a permission is a strict subset of another permission.
         .. describe:: x > y
 
+            Adds two permissions together.
```
Comment (EmreTech) [inline discord/permissions.py:100]: See above

**Review by EmreTech — COMMENTED** [review]
Looks good otherwise

File: discord/flags.py (hunk 5 above)
Comment (Lulalaby) [reply discord/flags.py:180]: Same confusion here

File: discord/flags.py [hunk 13]
```diff
@@ -177,6 +216,18 @@ class SystemChannelFlags(BaseFlags):
                Returns an iterator of ``(name, value)`` pairs. This allows it
                to be, for example, constructed as a dict or a list of pairs.
 
+        .. describe:: x + y
```
Comment (EmreTech) [inline discord/flags.py:219]: Newline + indent for the description

File: discord/flags.py [hunk 14]
```diff
@@ -177,6 +216,18 @@ class SystemChannelFlags(BaseFlags):
...
+         Adds two flags together.
+
+        .. describe:: x - y
```
Comment (EmreTech) [inline discord/flags.py:222]: Newline + indent for the description

File: discord/flags.py [hunk 15]
```diff
@@ -177,6 +216,18 @@ class SystemChannelFlags(BaseFlags):
...
+        Subtracts two flags from each other.
+
+        .. describe:: x & y
```
Comment (EmreTech) [inline discord/flags.py:225]: Newline + indent for the description

File: discord/flags.py [hunk 16]
```diff
@@ -177,6 +216,18 @@ class SystemChannelFlags(BaseFlags):
...
+        Returns the intersection of two flags.
+
+        .. describe:: ~x
```
Comment (EmreTech) [inline discord/flags.py:228]: Newline + indent for the description

File: discord/flags.py [hunk 17]
```diff
@@ -250,6 +301,19 @@ class MessageFlags(BaseFlags):
                Returns an iterator of ``(name, value)`` pairs. This allows it
                to be, for example, constructed as a dict or a list of pairs.
 
+        .. describe:: x + y
```
Comment (EmreTech) [inline discord/flags.py:304]: Newline + indent for the description

File: discord/flags.py [hunk 18]
```diff
@@ -250,6 +301,19 @@ class MessageFlags(BaseFlags):
...
+         Adds two flags together.
+
+        .. describe:: x - y
```
Comment (EmreTech) [inline discord/flags.py:307]: Newline + indent for the description

File: discord/flags.py [hunk 19]
```diff
@@ -250,6 +301,19 @@ class MessageFlags(BaseFlags):
...
+        Subtracts two flags from each other.
+
+        .. describe:: x & y
```
Comment (EmreTech) [inline discord/flags.py:310]: Newline + indent for the description

File: discord/flags.py [hunk 20]
```diff
@@ -250,6 +301,19 @@ class MessageFlags(BaseFlags):
...
+        Returns the intersection of two flags.
+
+        .. describe:: ~x
```
Comment (EmreTech) [inline discord/flags.py:313]: Newline + indent for the description

File: discord/flags.py [hunk 21]
```diff
@@ -346,6 +410,19 @@ class PublicUserFlags(BaseFlags):
             to be, for example, constructed as a dict or a list of pairs.
             Note that aliases are not shown.
 
+        .. describe:: x + y
```
Comment (EmreTech) [inline discord/flags.py:413]: Newline + indent for the description

File: discord/flags.py [hunk 22]
```diff
@@ -346,6 +410,19 @@ class PublicUserFlags(BaseFlags):
...
+         Adds two flags together.
+
+        .. describe:: x - y
```
Comment (EmreTech) [inline discord/flags.py:416]: Newline + indent for the description

File: discord/flags.py [hunk 23]
```diff
@@ -346,6 +410,19 @@ class PublicUserFlags(BaseFlags):
...
+        Subtracts two flags from each other.
+
+        .. describe:: x & y
```
Comment (EmreTech) [inline discord/flags.py:419]: Newline + indent for the description

File: discord/flags.py [hunk 24]
```diff
@@ -346,6 +410,19 @@ class PublicUserFlags(BaseFlags):
...
+        Returns the intersection of two flags.
+
+        .. describe:: ~x
```
Comment (EmreTech) [inline discord/flags.py:422]: Newline + indent for the description

File: discord/flags.py [hunk 25]
```diff
@@ -490,6 +567,19 @@ class Intents(BaseFlags):
                Returns an iterator of ``(name, value)`` pairs. This allows it
                to be, for example, constructed as a dict or a list of pairs.
 
+        .. describe:: x + y
```
Comment (EmreTech) [inline discord/flags.py:570]: Newline + indent for the description

File: discord/flags.py [hunk 26]
```diff
@@ -490,6 +567,19 @@ class Intents(BaseFlags):
...
+         Adds two flags together.
+
+        .. describe:: x - y
```
Comment (EmreTech) [inline discord/flags.py:573]: Newline + indent for the description

File: discord/flags.py [hunk 27]
```diff
@@ -490,6 +567,19 @@ class Intents(BaseFlags):
...
+        Subtracts two flags from each other.
+
+        .. describe:: x & y
```
Comment (EmreTech) [inline discord/flags.py:576]: Newline + indent for the description

File: discord/flags.py [hunk 28]
```diff
@@ -490,6 +567,19 @@ class Intents(BaseFlags):
...
+        Returns the intersection of two flags.
+
+        .. describe:: ~x
```
Comment (EmreTech) [inline discord/flags.py:579]: Newline + indent for the description

File: discord/flags.py [hunk 29]
```diff
@@ -1030,6 +1120,18 @@ class MemberCacheFlags(BaseFlags):
                Returns an iterator of ``(name, value)`` pairs. This allows it
                to be, for example, constructed as a dict or a list of pairs.
 
+        .. describe:: x + y
```
Comment (EmreTech) [inline discord/flags.py:1123]: Why is this here for the documentation of a specific flag

File: discord/flags.py [hunk 30]
```diff
@@ -1155,6 +1257,18 @@ class ApplicationFlags(BaseFlags):
             to be, for example, constructed as a dict or a list of pairs.
             Note that aliases are not shown.
 
+        .. describe:: x + y
```
Comment (EmreTech) [inline discord/flags.py:1260]: Newline + indent for the description

File: discord/flags.py [hunk 31]
```diff
@@ -1155,6 +1257,18 @@ class ApplicationFlags(BaseFlags):
...
+         Adds two flags together.
+
+        .. describe:: x - y
```
Comment (EmreTech) [inline discord/flags.py:1263]: Newline + indent for the description

File: discord/flags.py [hunk 32]
```diff
@@ -1155,6 +1257,18 @@ class ApplicationFlags(BaseFlags):
...
+        Subtracts two flags from each other.
+
+        .. describe:: x & y
```
Comment (EmreTech) [inline discord/flags.py:1266]: Newline + indent for the description

File: discord/flags.py [hunk 33]
```diff
@@ -1155,6 +1257,18 @@ class ApplicationFlags(BaseFlags):
...
+        Returns the intersection of two flags.
+
+        .. describe:: ~x
```
Comment (EmreTech) [inline discord/flags.py:1269]: Newline + indent for the description

File: discord/flags.py [hunk 34]
```diff
@@ -1253,6 +1367,18 @@ class ChannelFlags(BaseFlags):
             to be, for example, constructed as a dict or a list of pairs.
             Note that aliases are not shown.
 
+        .. describe:: x + y
```
Comment (EmreTech) [inline discord/flags.py:1370]: Newline + indent for the description

File: discord/flags.py [hunk 35]
```diff
@@ -1253,6 +1367,18 @@ class ChannelFlags(BaseFlags):
...
+         Adds two flags together.
+
+        .. describe:: x - y
```
Comment (EmreTech) [inline discord/flags.py:1373]: Newline + indent for the description

File: discord/flags.py [hunk 36]
```diff
@@ -1253,6 +1367,18 @@ class ChannelFlags(BaseFlags):
...
+        Subtracts two flags from each other.
+
+        .. describe:: x & y
```
Comment (EmreTech) [inline discord/flags.py:1376]: Newline + indent for the description

File: discord/flags.py [hunk 37]
```diff
@@ -1253,6 +1367,18 @@ class ChannelFlags(BaseFlags):
...
+        Returns the intersection of two flags.
+
+        .. describe:: ~x
```
Comment (EmreTech) [inline discord/flags.py:1379]: Newline + indent for the description

File: discord/permissions.py [hunk 38]
```diff
@@ -97,6 +97,18 @@ class Permissions(BaseFlags):
              Checks if a permission is a strict subset of another permission.
         .. describe:: x > y
 
+         .. describe:: x + y
```
Comment (EmreTech) [inline discord/permissions.py:100]: Newline + indent for the description

File: discord/permissions.py [hunk 39]
```diff
@@ -97,6 +97,18 @@ class Permissions(BaseFlags):
...
+         Adds two permissions together.
+
+        .. describe:: x - y
```
Comment (EmreTech) [inline discord/permissions.py:103]: Newline + indent for the description

File: discord/permissions.py [hunk 40]
```diff
@@ -97,6 +97,18 @@ class Permissions(BaseFlags):
...
+        Subtracts two permissions from each other.
+
+        .. describe:: x & y
```
Comment (EmreTech) [inline discord/permissions.py:106]: Newline + indent for the description

File: discord/permissions.py [hunk 41]
```diff
@@ -97,6 +97,18 @@ class Permissions(BaseFlags):
...
+        Returns the intersection of two permissions.
+
+        .. describe:: ~x
```
Comment (EmreTech) [inline discord/permissions.py:109]: Newline + indent for the description

File: discord/flags.py [hunk 42]
```diff
@@ -140,6 +141,44 @@ def __iter__(self) -> Iterator[Tuple[str, bool]]:
...
 
+    def __and__(self, other):
+        if isinstance(other, self.__class__):
+            return self.__class__._from_value(self.value & other.value)
+        elif isinstance(other, flag_value):
+            return self.__class__._from_value(self.value & other.flag)
+        else:
+            raise TypeError(f"'&' not supported between instances of {type(self)} and {type(other)}")
+
+    def __or__(self, other):
```
Comment (EmreTech) [inline discord/flags.py:152]: This doesn't seem to be documented anywhere

File: discord/flags.py (hunk 14 above)
Comment (Lulalaby) [inline discord/flags.py:222]:
```suggestion
        .. describe:: x + y
        Adds two flags together.

        .. describe:: x - y
```

File: discord/flags.py [hunk 43]
```diff
@@ -250,6 +301,19 @@ class MessageFlags(BaseFlags):
                Returns an iterator of ``(name, value)`` pairs. This allows it
                to be, for example, constructed as a dict or a list of pairs.
 
+        .. describe:: x + y
+         Adds two flags together.
```
Comment (Lulalaby) [inline discord/flags.py:305]:
```suggestion
        Adds two flags together.
```

**Review by EmreTech — COMMENTED** [review]
Sorry for the repetition, but it looks fine otherwise

File: discord/flags.py [hunk 44]
```diff
@@ -346,6 +410,19 @@ class PublicUserFlags(BaseFlags):
             to be, for example, constructed as a dict or a list of pairs.
             Note that aliases are not shown.
 
+        .. describe:: x + y
+         Adds two flags together.
```
Comment (Lulalaby) [inline discord/flags.py:414]:
```suggestion
        Adds two flags together.
```

File: discord/flags.py [hunk 45]
```diff
@@ -490,6 +567,19 @@ class Intents(BaseFlags):
                Returns an iterator of ``(name, value)`` pairs. This allows it
                to be, for example, constructed as a dict or a list of pairs.
 
+        .. describe:: x + y
+         Adds two flags together.
```
Comment (Lulalaby) [inline discord/flags.py:571]:
```suggestion
        Adds two flags together.
```

File: discord/flags.py [hunk 46]
```diff
@@ -1030,6 +1120,18 @@ class MemberCacheFlags(BaseFlags):
...
```
Comment (Lulalaby) [inline discord/flags.py:1124]:
```suggestion
        Adds two flags together.
```

File: discord/flags.py [hunk 47]
```diff
@@ -1155,6 +1257,18 @@ class ApplicationFlags(BaseFlags):
             to be, for example, constructed as a dict or a list of pairs.
             Note that aliases are not shown.
 
+        .. describe:: x + y
+         Adds two flags together.
```
Comment (Lulalaby) [inline discord/flags.py:1261]:
```suggestion
        Adds two flags together.
```

File: discord/flags.py [hunk 48]
```diff
@@ -1253,6 +1367,18 @@ class ChannelFlags(BaseFlags):
...
```
Comment (Lulalaby) [inline discord/flags.py:1371]:
```suggestion
        Adds two flags together.
```

File: discord/permissions.py (hunk 39 above)
Comment (Lulalaby) [inline discord/permissions.py:103]:
```suggestion
        .. describe:: x + y
        Adds two permissions together.

        .. describe:: x - y
```

**Review by Lulalaby — DISMISSED** [review]

File: discord/flags.py (hunk 29 above)
Comment (plun1331) [reply discord/flags.py:1123]: It's under `MemberCacheFlags`

Comment (Lulalaby) [reply discord/flags.py:1123]: Shouldn't _MemberCacheFlags(BaseFlags)_ get the operators / methods from base flags already?

Comment (plun1331) [reply discord/flags.py:1123]: > Shouldn't _MemberCacheFlags(BaseFlags)_ get the operators / methods from base flags already?

It'll get them, they just won't be in the documentation unless they're put there

Comment (Lulalaby) [reply discord/flags.py:1123]: Wouldn't autodoc be a good approach? Or doesn't it work.

Comment (plun1331) [reply discord/flags.py:1123]: autodoc is for detecting modules/classes and their methods and automatically adding them to the documentation, it won't merge docstrings from two different classes

Comment (Lulalaby) [reply discord/flags.py:1123]: What the.. that sounds so useless. 
Well ok fair.

**Review by Middledot — APPROVED** [review]

**Review by Lulalaby — APPROVED** [review]
